File: code/zato-web-admin/src/zato/admin/static/js/cloud/dropbox.js

// /////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.Dropbox = new Class({
    toString: function() {
        var s = '<Dropbox id:{0} name:{1} is_active:{2}';
        return String.format(s, this.id ? this.id : '(none)',
                                this.name ? this.name : '(none)',
                                this.is_active ? this.is_active : '(none)');
    }
});

// /////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.password_required = false;
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.Dropbox;
    $.fn.zato.data_table.new_row_func = $.fn.zato.cloud.dropbox.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms(['name', 'timeout', 'user_agent']);
})

$.fn.zato.cloud.dropbox.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new Dropbox connection', null);
}

$.fn.zato.cloud.dropbox.edit = function(id) {
    $.fn.zato.data_table._create_edit('edit', 'Update the Dropbox connection', id);
}

$.fn.zato.cloud.dropbox.data_table.new_row = function(item, data, include_tr) {
    var row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    var is_active = item.is_active == true;

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";

    // 1
    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', is_active ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', item.user_agent);

    // 2
    row += String.format('<td>{0}</td>', item.default_scope || $.fn.zato.empty_value);
    row += String.format('<td>{0}</td>', item.default_directory || $.fn.zato.empty_value);

    // 3
    row += String.format('<td>{0}</td>',
        String.format("<a href=\"javascript:$.fn.zato.data_table.change_password('{0}', 'Change token')\">Change token</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.cloud.dropbox.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.cloud.dropbox.delete_({0});'>Delete</a>", item.id));

    // 4
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.data_table.ping({0});'>Ping</a>", item.id));
    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);
    row += String.format("<td class='ignore'>{0}</td>", item.max_retries_on_error || '');

    // 5
    row += String.format("<td class='ignore'>{0}</td>", item.max_retries_on_rate_limit || '');
    row += String.format("<td class='ignore'>{0}</td>", item.timeout || '');
    row += String.format("<td class='ignore'>{0}</td>", item.oauth2_access_token_expiration || '');

    // 6
    row += String.format("<td class='ignore'>{0}</td>", item.default_scope || '');
    row += String.format("<td class='ignore'>{0}</td>", item.pool_size);

    // 7
    row += String.format("<td class='ignore'>{0}</td>", item.default_directory || '');
    row += String.format("<td class='ignore'>{0}</td>", item.http_headers || '');
    row += String.format("<td class='ignore'>{0}</td>", item.is_active);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

$.fn.zato.cloud.dropbox.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'Outgoing Dropbox connection `{0}` deleted',
        'Are you sure you want to delete Dropbox connection `{0}`?',
        true);
}

File: code/zato-web-admin/src/zato/admin/web/views/cloud/dropbox.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2022, Zato Source s.r.o. https://zato.io

Licensed under AGPLv3, see LICENSE.txt for terms and conditions.
"""

# stdlib
import logging

# Zato
from zato.common.api import GENERIC
from zato.admin.web.forms import ChangePasswordForm
from zato.admin.web.forms.cloud.dropbox import CreateForm, EditForm
from zato.admin.web.views import change_password as _change_password, CreateEdit, Delete as _Delete, Index as _Index, \
     method_allowed, ping_connection, slugify
from zato.common.odb.model import GenericConn

# ################################################################################################################################

logger = logging.getLogger(__name__)

# ################################################################################################################################

_sio_required = 'name',
_sio_optional = 'is_active', 'max_retries_on_error', 'max_retries_on_rate_limit', 'timeout', 'oauth2_access_token_expiration', \
    'oauth2_access_token', 'default_directory', 'default_scope', 'user_agent', 'http_headers'

# ################################################################################################################################
# ################################################################################################################################

class Index(_Index):
    method_allowed = 'GET'
    url_name = 'cloud-dropbox'
    template = 'zato/cloud/dropbox.html'
    service_name = 'zato.generic.connection.get-list'
    output_class = GenericConn
    paginate = True

    class SimpleIO(_Index.SimpleIO):
        input_required = ('cluster_id', 'type_')
        output_required = ('id',) + _sio_required
        output_optional = _sio_optional
        output_repeated = True

    def handle(self):
        return {
            'create_form': CreateForm(req=self.req),
            'edit_form': EditForm(prefix='edit', req=self.req),
            'change_password_form': ChangePasswordForm()
        }

# ################################################################################################################################
# ################################################################################################################################

class _CreateEdit(CreateEdit):
    method_allowed = 'POST'

    class SimpleIO(CreateEdit.SimpleIO):
        input_required = _sio_required + _sio_optional
        output_required = ('id', 'name')

    def populate_initial_input_dict(self, initial_input_dict):
        initial_input_dict['type_'] = GENERIC.CONNECTION.TYPE.CLOUD_DROPBOX
        initial_input_dict['is_internal'] = False
        initial_input_dict['is_channel'] = False
        initial_input_dict['is_outconn'] = True
        initial_input_dict['pool_size'] = 10
        initial_input_dict['sec_use_rbac'] = False

    def post_process_return_data(self, return_data):
        return_data['name_slug'] = slugify(return_data['name'])
        return return_data

    def success_message(self, item):
        return 'Successfully {} Dropbox connection `{}`'.format(self.verb, item.name)

# ################################################################################################################################
# ################################################################################################################################

class Create(_CreateEdit):
    url_name = 'cloud-dropbox-create'
    service_name = 'zato.generic.connection.create'

# ################################################################################################################################
# ################################################################################################################################

class Edit(_CreateEdit):
    url_name = 'cloud-dropbox-edit'
    form_prefix = 'edit-'
    service_name = 'zato.generic.connection.edit'

# ################################################################################################################################
# ################################################################################################################################

class Delete(_Delete):
    url_name = 'cloud-dropbox-delete'
    error_message = 'Could not delete Dropbox connection'
    service_name = 'zato.generic.connection.delete'

# ################################################################################################################################
# ################################################################################################################################

@method_allowed('POST')
def change_password(req):
    return _change_password(req, 'zato.generic.connection.change-password', success_msg='Token updated')

# ################################################################################################################################

@method_allowed('POST')
def ping(req, id, cluster_id):
    return ping_connection(req, 'zato.generic.connection.ping', id, 'Dropbox connection')

# ################################################################################################################################
